### PR TITLE
Ffi speed

### PR DESCRIPTION
**Pull request by bgrieder**
- added FFI hybrid encryption using a cache to speed up encryption/decryption (x4 on encryption, x2 on decryption)
- added FFI tests
- added benchmarks and profiling- see `bench.rs` for details

Run benchmarks:

```sh
cargo run --release --features ffi --bin bench_abe_gpsw
```

### CODE REVIEW

File: src/bench.rs [hunk 1]
```diff
@@ -192,22 +259,290 @@ pub unsafe fn bench_ffi_header_encryption() -> anyhow::Result<()> {
...
+    use abe_gpsw::interfaces::ffi::hybrid_gpsw_aes::{
+        h_aes_create_encryption_cache, h_aes_destroy_encryption_cache,
+        h_aes_encrypt_header_using_cache,
+    };
+    use serde_json::Value;
+    use std::{
+        ffi::{c_void, CString},
+        os::raw::{c_char, c_int},
+        time::Instant,
+    };
```
Comment (ThibsG) [inline src/bench.rs:286]: imports at top of the file

File: src/bench.rs [hunk 2]
```diff
@@ -192,22 +259,290 @@ pub unsafe fn bench_ffi_header_encryption() -> anyhow::Result<()> {
...
+    unwrap_ffi_error(h_aes_destroy_encryption_cache(cache_ptr_ptr))?;
     Ok(())
 }
+
+///
+/// # Safety
+#[cfg(feature = "interfaces")]
+pub fn bench_header_decryption() -> anyhow::Result<()> {
+    use serde_json::Value;
+    use std::time::Instant;
```
Comment (ThibsG) [inline src/bench.rs:370]: imports at top of the file

File: src/bench.rs [hunk 3]
```diff
@@ -192,22 +259,290 @@ pub unsafe fn bench_ffi_header_encryption() -> anyhow::Result<()> {
...
+///
+/// # Safety
+#[cfg(feature = "ffi")]
+pub unsafe fn bench_ffi_header_decryption() -> anyhow::Result<()> {
+    use abe_gpsw::interfaces::ffi::hybrid_gpsw_aes::h_aes_decrypt_header;
+    use serde_json::Value;
+    use std::{
+        os::raw::{c_char, c_int},
+        time::Instant,
+    };
```
Comment (ThibsG) [inline src/bench.rs:405]: imports at top of the file (more below)

**Review by ThibsG — CHANGES_REQUESTED** [review]
on a general note, is there any blocker to use `cargo bench` in this case?

File: src/bench.rs [hunk 4]
```diff
@@ -55,26 +45,57 @@ fn main() -> anyhow::Result<()> {
...
     } else if selector == "all" {
         #[cfg(feature = "interfaces")]
         bench_header_encryption()?;
-
-        #[cfg(feature = "ffi")]
         unsafe {
             bench_ffi_header_encryption()?;
+            #[cfg(feature = "ffi")]
+            bench_ffi_header_encryption_using_cache()?;
+        }
```
[the remark this answers is not in the document]
Comment (ThibsG) [reply src/bench.rs:72]: btw, with this change, you can remove test only on libs, code will compile again

File: src/bench.rs [hunk 5]
```diff
@@ -3,38 +3,29 @@
...
-#[cfg(feature = "ffi")]
-use {
-    abe_gpsw::interfaces::ffi::{error::get_last_error, hybrid_gpsw_aes::h_aes_encrypt_header},
-    std::{
-        ffi::{CStr, CString},
-        os::raw::{c_char, c_int},
-    },
+#[cfg(feature = "interfaces")]
+use cosmian_crypto_base::{
+    hybrid_crypto::Metadata, symmetric_crypto::aes_256_gcm_pure::Aes256GcmCrypto,
```
[the remark this answers is not in the document]
Comment (bgrieder) [reply src/bench.rs:19]: Done

File: src/bench.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (bgrieder) [reply src/bench.rs:72]: done

File: src/bench.rs [hunk 6]
```diff
@@ -55,26 +45,57 @@ fn main() -> anyhow::Result<()> {
...
+            bench_ffi_header_encryption_using_cache()?;
+        }
+        #[cfg(feature = "interfaces")]
+        bench_header_decryption()?;
+        unsafe {
+            #[cfg(feature = "ffi")]
+            bench_ffi_header_decryption()?;
+            #[cfg(feature = "ffi")]
+            bench_ffi_header_decryption_using_cache()?;
         }
```
[the remark this answers is not in the document]
Comment (bgrieder) [reply src/bench.rs:80]: done

File: src/bench.rs [hunk 7]
```diff
@@ -92,6 +113,9 @@ see above for the OPTION values
 
 #[cfg(feature = "interfaces")]
 pub fn bench_header_encryption() -> anyhow::Result<()> {
+    use serde_json::Value;
+    use std::time::Instant;
```
[the remark this answers is not in the document]
Comment (bgrieder) [reply src/bench.rs:116]: done

File: src/bench.rs [hunk 8]
```diff
@@ -132,10 +156,53 @@ pub fn bench_header_encryption() -> anyhow::Result<()> {
...
 /// # Safety
 #[cfg(feature = "ffi")]
 pub unsafe fn bench_ffi_header_encryption() -> anyhow::Result<()> {
+    use abe_gpsw::interfaces::ffi::hybrid_gpsw_aes::h_aes_encrypt_header;
+    use serde_json::Value;
+    use std::{
+        ffi::CString,
+        os::raw::{c_char, c_int},
+        time::Instant,
+    };
```
[the remark this answers is not in the document]
Comment (bgrieder) [reply src/bench.rs:204]: done

File: src/bench.rs (hunk 1 above)
Comment (bgrieder) [reply src/bench.rs:286]: done

File: src/bench.rs (hunk 2 above)
Comment (bgrieder) [reply src/bench.rs:370]: done

File: src/bench.rs (hunk 3 above)
Comment (bgrieder) [reply src/bench.rs:405]: done

**Comment by ThibsG** [comment]
can you also activate back testing everything (remove `--lib` here: https://github.com/Cosmian/abe_gpsw/pull/17/commits/c16c542b994a16633d5af007fe2623f6e74d5c61#diff-b803fcb7f17ed9235f1e5cb1fcd2f5d3b2838429d4368ae4c57ce4436577f03fR54)?

**Comment by ThibsG** [comment]
about CI failing: `serde_json::Value` and `std::time::Instant` should also be present under `interfaces` feature also

**Comment by bgrieder** [comment]
@ThibsG  This creates constant issues with the WASM target
All the tests are in the lib anyway. So let us leave it like this way
If you feel like fixing this, I let you open an issue/PR :)

**Pull request merged**